components/service/glean/src/main/java/mozilla/components/service/glean/storages/UuidsStorageEngine.kt

```diff
@@ -4,17 +4,47 @@
 
 package mozilla.components.service.glean.storages
 
+import android.annotation.SuppressLint
+import android.content.Context
+import android.content.SharedPreferences
 import java.util.UUID
 
 import android.support.annotation.VisibleForTesting
+import mozilla.components.service.glean.Lifetime
+import mozilla.components.support.base.log.logger.Logger
+import org.json.JSONObject
+
+// Convenience aliases to make the code more readable.
+internal typealias UUIDStorage = MutableMap<String, UUID>
+internal typealias UUIDMap = MutableMap<String, UUIDStorage>
 
 /**
  * This singleton handles the in-memory storage logic for uuids. It is meant to be used by
  * the Specific UUID API and the ping assembling objects. No validation on the stored data
  * is performed at this point: validation must be performed by the Specific Uuids API.
+ *
+ * This class contains a reference to the Android application Context. While the IDE warns
+ * us that this could leak, the application context lives as long as the application and this
+ * object. For this reason, we should be safe to suppress the IDE warning.
+ */
+@SuppressLint("StaticFieldLeak")
+internal object UuidsStorageEngine : UuidsStorageEngineImplementation()
+
+/**
+ * This class implements the behaviour for UuidsStorageEngine. This is separate
+ * from the object to make it easier to test it.
  */
-internal object UuidsStorageEngine {
-    private val uuidStores: MutableMap<String, MutableMap<String, UUID>> = mutableMapOf()
+@VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+open class UuidsStorageEngineImplementation : StorageEngine {
+    override lateinit var applicationContext: Context
+
+    private val userLifetimeStorage: SharedPreferences by lazy { deserializeUserLifetime() }
+    private val logger = Logger("glean/HttpPingUploader")
+
+    // Use a multi-level map to store the data for the different lifetimes and
+    // stores: Map[Lifetime, Map[StorageName, UUID]].
+    private val uuidStores: Map<String, UUIDMap> =
```

Comment:
Since `Lifetime` is an enum which we know is a small number of contiguous integers, we can probably use `Array<UUIDMap>` here instead and save all of the hashing / toStringing of enumerations.